client/X11/xf_cliprdr.c

```diff
@@ -1663,6 +1928,334 @@ static UINT xf_cliprdr_clipboard_file_range_failure(wClipboardDelegate* delegate
 	return clipboard->context->ClientFileContentsResponse(clipboard->context, &response);
 }
 
+static void xf_cliprdr_fuse_getattr(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info* fi)
+{
+	double timeout = 0;
+	struct stat stbuf;
+
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+	else
+	{
+		memset(&stbuf, 0, sizeof(stbuf));
+		stbuf.st_ino = ino;
+		stbuf.st_mode = node->st_mode;
+		stbuf.st_mtime = node->st_mtim.tv_sec;
+		if (ino == FUSE_ROOT_ID)
+		{
+			stbuf.st_nlink = 2;
+		}
+		else
+		{
+			stbuf.st_nlink = 1;
+		}
+		fuse_reply_attr(req, &stbuf, timeout);
+	}
+}
+
+static void xf_cliprdr_fuse_readdir(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
+                                    struct fuse_file_info* fi)
+{
+	size_t count;
+	size_t index;
+	size_t child_ino;
+	size_t direntry_len;
+	char* buf = (char*)calloc(size, sizeof(char));
+	size_t pos = 0;
+	xfCliprdrFuseInode* child_node;
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node || !node->child_inos)
+	{
+		fuse_reply_err(req, ENOENT);
+		return;
+	}
+	else if ((node->st_mode & S_IFDIR) == 0)
+	{
+		fuse_reply_err(req, ENOTDIR);
+		return;
+	}
+
+	ArrayList_Lock(node->child_inos);
+	count = ArrayList_Count(node->child_inos);
+	if (count == 0 || count + 1 <= off)
+	{
+		fuse_reply_buf(req, NULL, 0);
+	}
+	else
+	{
+		for (index = off; index < count + 2; index++)
+		{
+			if (index == 0)
+			{
+				struct stat stbuf;
+				stbuf.st_ino = ino;
+				direntry_len = fuse_add_direntry(req, NULL, 0, ".", NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+				fuse_add_direntry(req, buf + pos, size - pos, ".", &stbuf, index);
```

Comment:
I think it is enough to do only one call here:
```c
direntry_len = fuse_add_direntry(req, buf + pos, size - pos, ".", &stbuf, index);
if (direntry_len > size - pos)
	break;
```
since `fuse_add_direntry()` promises to not modify the buffer if it cannot fit an entry there, but will still return the expected length.

Ditto for other call sites.